src/_articles/libraries/dart-io.md

```diff
@@ -0,0 +1,386 @@
+and hook up a listener (using `await for`) to its stream of `HttpRequest`s.
+
+Here is a simple web server
+that just answers 'Hello, world' to any request.
+
+<?code-excerpt "misc/lib/articles/io/io_http_server_test.dart"?>
+{% prettify dart %}
+import 'dart:async' show Future;
+import 'dart:io';
+
+Future<void> main() async {
+  HttpServer server = await HttpServer.bind('127.0.0.1', 8082);
+  await for (HttpRequest request in server) {
+    request.response.write('Hello, world');
+    request.response.close();
+  }
+}
+{% endprettify %}
+
+Running this application
+and pointing your browser to 'http://127.0.0.1:8082'
+gives you 'Hello, world' as expected.
+
+Let's add a bit more and actually serve files.
+The base path for every file that we serve will be
+the location of the script.
+If no path is specified in a request we will serve index.html.
+For a request with a path,
+we will attempt to find the file and serve it.
+If the file is not found we will respond with a '404 Not Found' status.
+We make use of the streaming interface
+to pipe all the data read from a file directly to the response stream.
+
+<?code-excerpt "misc/lib/articles/io/io_http_server_file_test.dart"?>
+{% prettify dart %}
+import 'dart:async' show Future;
+import 'dart:io';
+
+Future<void> runServer(String basePath) async {
+  HttpServer server = await HttpServer.bind('127.0.0.1', 8082);
+  await for (HttpRequest request in server) {
+    handleRequest(basePath, request);
+  }
+}
+
+Future<void> handleRequest(String basePath, HttpRequest request) async {
+  final String path = request.uri.toFilePath();
+  // PENDING: Do more security checks here.
+  final String resultPath = path == '/' ? '/index.html' : path;
+  final File file = File('$basePath$resultPath');
+  if (await file.exists()) {
+    try {
+      await file.openRead().pipe(request.response);
+    } catch (exception) {
+      print('Error happened: $exception');
+      await sendInternalError(request.response);
+    }
+  } else {
+    await sendNotFound(request.response);
+  }
+}
+
+Future<void> sendInternalError(HttpResponse response) async {
+  response.statusCode = HttpStatus.internalServerError;
+  await response.close();
+}
+
+Future<void> sendNotFound(HttpResponse response) async {
+  response.statusCode = HttpStatus.notFound;
+  await response.close();
+}
+
+Future<void> main() async {
+  // Compute base path for the request based on the location of the
+  // script, and then start the server.
+  final script = File(Platform.script.toFilePath());
+  await runServer(script.parent.path);
+}
+{% endprettify %}
+
+Writing HTTP clients is very similar to using the
+[HttpClient]({{site.dart_api}}/{{site.data.pkg-vers.SDK.channel}}/dart-io/HttpClient-class.html)
+class.
+
+
+## Feature requests welcome
+
+The dart:io library is already capable of performing a lot of tasks.
+As an example of our own use of dart:io,
+we have rewritten the Dart testing scripts from Python to Dart.
```

Comment:
I think that our testing scripts were originally written in Python might not be very interesting. Maybe we could change this to:

The dart:io library is capable of performing a lot of tasks. As an example the `pub.dartlang.org` site is written using it.